### PR TITLE
Added select_one_from file and select_multiple_from_file, close #164

### LINKED ISSUE #164: Document select_one_from_file / select_multiple_from_file
https://github.com/XLSForm/pyxform/issues/30

### PR DESCRIPTION
**Pull request by MartijnR**

### CODE REVIEW

**Comment by lognaturel** [comment]
I feel like we should at least make some attempt of acknowledging that there's a connection between this and the external XML data section and contrasting it with the "Dynamic selects from pre-loaded data" and "External selects" sections. What do you think?

**Comment by MartijnR** [comment]
Yes, and vice versa. Good point. Will try something and steer users to these 'proper XForm' solutions if they work for them.

**Comment by MartijnR** [comment]
Forgot to commit this work. Let me know what you think (though no hurry)!

**Comment by lognaturel** [comment]
Sorry about letting this sit. I have all kinds of thoughts. Would you be up for a 30min call to discuss? I think that might be more efficient than going back and forth as I have a number of questions for you about Enketo performance and external datasets in general. Let me know if so and I'll reach out to schedule.

**Comment by MartijnR** [comment]
> I have all kinds of thoughts.
Uh oh.... :D 

Thanks, yes, I will ping you in the next few days to find a time. Now getting ready to battle a snow storm (and play in the snow).

**Comment by lognaturel** [comment]
> Uh oh.... :D

I know! 😫

> Now getting ready to battle a snow storm (and play in the snow).

Hard to imagine from here -- "we put the 'sweat' in 'sweater weather.'" Hope it all goes well and there's some fun to be had! ❄️

**Comment by MartijnR** [comment]
I am quite curious now to try to load a form in Enketo with hundreds of thousand of data items that provides the rationale for select_external in ODK Collect. Do you perhaps have a real-life form for which select_external is required?

If it does load well (who knows, but probably not), it's worth exploring a transformation of _select_external_ in enketo-transformer into regular external CSV data syntax (with a workaround for translations).

**Comment by lognaturel** [comment]
[nigeria_wards_internal_simplified.xml.zip](https://github.com/XLSForm/xlsform.github.io/files/3806397/nigeria_wards_internal_simplified.xml.zip) is an example we discussed privately with ~10k items. This kind of form was the rationale for the features originally. It now performs reasonably well on modern devices and with performance improvements made over the last couple of years. Sadly, it doesn't look like there was any attempt to speed up the spec-compliant implementation before adding the database-backed implementations.

Unfortunately, I don't have access to real data that gets to larger scales. I'll see if I can get my hands on some or find a real deployment whose data we can mimic.

File: _sections/home-english.md [hunk 1]
```diff
@@ -172,6 +174,35 @@ Click on the link to look at the complete [pizza_questionnaire](https://docs.goo
 **Caveat**  
 When you export data using this **or_other** option, in the **favorite_topping** column, you will see a value **other**. A separate column will have the answer for the questions in which the user selected **other**. This makes data analysis more cumbersome, so we do not recommend the **or_other** construct for large scale data collection efforts. See the **Relevant** section below for an alternative method more appropriate for large scale projects.
 
+### Multiple choice from file
+
+The options in a multiple choice question can also be taken from a separate file instead of the choices sheet. This is particularly useful if the options are dynamic or if the list of options is used in multiple surveys. Two types of files are supported: CSV and XML files. See usage examples below:
+
+| type                                    | name | label                          | choice_filter   |
+| --------------------------------------- | ---- | ------------------------------ | --------------- |
+| select_multiple_from_file country.csv   | dest | Which countries did you live?  |                 |
```
Comment (lognaturel) [inline _sections/home-english.md]: Should probably be "**In** which countries did you live?" or "Which countries did you live **in**?". Does `dest` make sense as a name?

File: _sections/home-english.md [hunk 2]
```diff
@@ -172,6 +174,35 @@ Click on the link to look at the complete [pizza_questionnaire](https://docs.goo
...
+| select_one_from_file countries.xml      | cou  | Which country do you live now? |                 |
```
Comment (lognaturel) [inline _sections/home-english.md]: Same as above -- "**In** which country do you live now?" or "Which country do you live **in** now?"

**Comment by MartijnR** [comment]
> It now performs reasonably well on modern devices and with performance improvements made over the last couple of years. 

Oh, that's great. I misunderstood you during our last meeting. So the path is open to promoting these new spec-compliant ways.

File: _sections/home-english.md (hunk 1 above)
Comment (MartijnR) [reply _sections/home-english.md]: Good point. Will change! Thx

**Comment by MartijnR** [comment]
> I'm ok with this. I expect there will be some confusion because it comes before the other options but it's probably not a big deal and we can adjust as needed.

Yes, I agree, external data is messy, reflecting implementations ;). Indeed, we can improve it later.

**Comment by lognaturel** [comment]
> So the path is open to promoting these new spec-compliant ways.

I think so? Like you, I think I need to see examples of truly large datasets to see how they're performing these days.

I think another thing that's missing is documentation on the full workflow from "I have a CSV that was created by some other process" to "my form definition queries that data and uses it in selects".

It's unfortunate that columns named `name` and `label` are required for selects. We might look at making those configurable in XLSForm at some point.

For the actual querying, I go back and forth on how that's best done. As much as I don't love `pulldata` and find it personally confusing, perhaps it's the best option. I think we should have more resources and XLSForm support for more complex XPath but I don't know how reasonable it is for users to make deep use of (see [this thread](https://forum.opendatakit.org/t/xlsform-call-a-column-from-choices-in-the-relevant-field/22972/6?u=ln) for a recent example).

**Pull request merged**